### PR TITLE
Simplify OpenMP detection in CMake

### PR DESCRIPTION
**Pull request by cyyever**
We greatly simplify the handing of OpenMP in CMake by using caffe2::openmp target thoroughly. We follow the old behavior by defaulting to MKL OMP library and detecting OMP flags otherwise.

cc @malfet @seemethere

### CODE REVIEW

File: cmake/public/cuda.cmake [hunk 1]
```diff
@@ -461,17 +461,6 @@ if(MSVC)
...
-if("${CMAKE_CXX_SIMULATE_ID}" STREQUAL "MSVC"
-  AND "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-  list(APPEND CUDA_PROPAGATE_HOST_FLAGS_BLOCKLIST "-Xclang" "-fopenmp")
-  if(MSVC_TOOLSET_VERSION LESS 142)
-    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-openmp")
-  else()
-    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-openmp:experimental")
-  endif()
-endif()
-
```
Comment (malfet) [inline cmake/public/cuda.cmake:474]: Hmm, did you test this codepath? (Not that we use it frequently, but still why remove it?)

**Review by malfet — COMMENTED** [review]
Test regressions seem relevant, aren't they?

File: cmake/public/cuda.cmake (hunk 1 above)
Comment (cyyever) [reply cmake/public/cuda.cmake:474]: here it uses CUDA_PROPAGATE_HOST_FLAGS_BLOCKLIST, but flags are appended to  CUDA_PROPAGATE_HOST_FLAGS_BLACKLIST, so this code has no effect. I think we should handle host code omp flags forward in openmp target?

**Comment by cyyever** [comment]
> Test regressions seem relevant, aren't they?

Most of them failed in test_torchinductor. I have to rebase  and rerun them when the master become green.

**Comment by cyyever** [comment]
@malfet Clang-7 failed with Openmp, but clang-7 is too old, I can't setup  a local environment to reproduce it easily...

**Comment by cyyever** [comment]
@malfet  Can you review this PR? I have managed to pass all tests.

**Review by malfet — APPROVED** [review]
LGTM, thank you for the improvements

**Comment by cyyever** [comment]
@malfet Thank you for acceptance, I am preparing more CMake refactoring

**Comment by cyyever** [comment]
@malfet The failed case export USE_CUPTI_SO=0 in CI script. But CMake requires USE_CUPTI_SO=1, we should change the CI script

**Comment by malfet** [comment]
@cyyever it comes from https://github.com/pytorch/builder/blob/16fb06efc2b73abacf7b0c55c20632ec28919e98/manywheel/build_cuda.sh#L15
but I don't think it's necessary to change that invocation, but rather understand, what have change in the build system that caused a regression. Likely some of the flags that were set before the change are no longer there, so please investigate what have changed. (and whether there are any suspicious binary size changes)

**Review by malfet — APPROVED** [review]

**Comment by cyyever** [comment]
@malfet  After several hours of investigation, I can't reproduce the error. Then I suspect it is a CentOS 7 specific issue. Using the builder scripts, I encountered some yum mirror errors. Given that CentOS 7 is going to dead, I think it pay little to try to fix it...

**Comment by cyyever** [comment]
@malfet USE_CUPTI_SO=0 and centos7 combination needs to link pthread to run the cmake test. I have added it.

**Pull request closed without merging**